contracts/starknet/end.cairo

```diff
@@ -0,0 +1,934 @@
+    auth();
+    require_live();
+
+    // live = 0;
+    _live.write(0);
+
+    let (timestamp) = get_block_timestamp();
+    _when.write(timestamp);
+
+    let (vat) = _vat.read();
+    let (spot) = _spot.read();
+    let (pot) = _pot.read();
+    let (cure) = _cure.read();
+
+    VatLike.cage(vat);
+    SpotLike.cage(spot);
+    PotLike.cage(pot);
+    CureLike.cage(cure);
+
+    // emit Cage();
+    Cage.emit();
+
+    return ();
+}
+
+// function cage(bytes32 ilk) external {
+@external
+func cage_ilk{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(ilk: felt) {
+    alloc_locals;
+    // require(live == 0, "End/still-live");
+    require_live();
+    // require(tag[ilk] == 0, "End/tag-ilk-already-defined");
+    let (tag) = _tag.read(ilk);
+    with_attr error_message("End/tag-ilk-already-defined") {
+        assert_0(tag);
+    }
+    // (Art[ilk],,,,) = vat.ilks(ilk);
+    let (vat) = _vat.read();
+    let (Art: Uint256, _, _, _, _) = VatLike.ilks(vat, ilk);
+    _Art.write(ilk, Art);
+    // (PipLike pip,) = spot.ilks(ilk);
+    let (spot) = _spot.read();
+    let (pip: felt, _) = SpotLike.ilks(spot, ilk);
+    // // par is a ray, pip returns a wad
+    let (par) = SpotLike.par(spot);
+    let (pip_val) = PipLike.read(pip);
+    let (local pip_val_uint) = _felt_to_uint(pip_val);
+    let (local pip_val_ray: Ray) = wad_to_ray(Wad(pip_val_uint));
+    // tag[ilk] = wdiv(spot.par(), uint256(pip.read()));
+    let (new_tag) = ray_div(Ray(par), pip_val_ray);
+    _tag.write(ilk, new_tag.ray);
+    // emit Cage(ilk);
+    Cage_ilk.emit(ilk);
+    // }
+
+    return ();
+}
+
+// function skim(bytes32 ilk, address urn) external {
+@external
+func skim{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
+}(ilk: felt, urn: felt) {
+    alloc_locals;
+    // require(tag[ilk] != 0, "End/tag-ilk-not-defined");
+    let (tag) = _tag.read(ilk);
+    with_attr error_message("End/tag-ilk-not-defined") {
+        assert_not_0(tag);
+    }
+    // (, uint256 rate,,,) = vat.ilks(ilk);
+    let (vat) = _vat.read();
+    let (_, rate: Uint256, _, _, _) = VatLike.ilks(vat, ilk);
+    // (uint256 ink, uint256 art) = vat.urns(ilk, urn);
+    let (ink: Uint256, art: Uint256) = VatLike.urns(vat, ilk, urn);
+
+    // uint256 owe = rmul(rmul(art, rate), tag[ilk]);
+    local art_: Ray = Ray(art);
+    local rate_: Ray = Ray(art);
+    let (mul: Ray) = ray_mul_no_rounding(art_, rate_);
+    local tag_: Ray = Ray(tag);
+    let (owe) = ray_mul_no_rounding(mul, tag_);
+    // uint256 wad = min(ink, owe);
+    let (wad) = _min(ink, owe.ray);
+    // gap[ilk] = gap[ilk] + (owe - wad);
+    let (gap) = _gap.read(ilk);
+    let (sub_) = sub(owe.ray, wad);
+    let (new_gap) = add(gap, sub_);
+    _gap.write(ilk, new_gap);
+
+    // require(wad <= 2**255 && art <= 2**255, "End/overflow");
+    with_attr error_message("End/overflow") {
+        let (max) = _felt_to_uint(2 ** 255);
```

Comment:
in felt: 2**255 = SN_PRIME - 1707371672025177167763334723112485265179840680894217387507728, it is not what you want

Reply:
Why is it resolved?